dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,397 @@
+import io.etcd.jetcd.options.PutOption;
+import io.etcd.jetcd.options.WatchOption;
+import io.etcd.jetcd.watch.WatchEvent;
+import io.etcd.jetcd.watch.WatchResponse;
+import io.netty.util.internal.StringUtil;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public class AlluxioEtcdClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final Lock INSTANCE_LOCK = new ReentrantLock();
+  @GuardedBy("INSTANCE_LOCK")
+  private static final AtomicReference<AlluxioEtcdClient> ALLUXIO_ETCD_CLIENT = new AtomicReference<>();
+  protected AtomicBoolean mConnected = new AtomicBoolean(false);
+  private Client mClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  public String[] mEndpoints = new String[0];
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[endpointsList.size()]);
+    mServiceDiscovery = new ServiceDiscoveryRecipe(this, clusterName);
+  }
+
+  public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
+    if (ALLUXIO_ETCD_CLIENT.get() == null) {
+      try (LockResource lockResource = new LockResource(INSTANCE_LOCK)) {
+        if (ALLUXIO_ETCD_CLIENT.get() == null) {
+          ALLUXIO_ETCD_CLIENT.set(new AlluxioEtcdClient(conf));
+        }
+      }
+    }
+    return ALLUXIO_ETCD_CLIENT.get();
+  }
+
+  public void connect() {
+    connect(false);
+  }
+
+  public void connect(boolean force) {
+    if (mConnected.get() && !force) {
+      return;
+    }
+    mConnected.set(false);
+    // create client using endpoints
+    Client client = Client.builder().endpoints(mEndpoints)
+        .build();
+    if (mConnected.compareAndSet(false, true)) {
+      mClient = client;
+    }
+  }
+
```

Comment:
Client creation should equal client connection, as is the case for gRPC clients and raw etcd clients. wdyt?

Reply:
i actually refer to AbstractClient.connect() which is called within retryRpcInternal, and got triggered on first rpc call made to grpc services. Therefore AlluxioEtcdClient.connect() is getting called on first usage of itself, like getEtcdClient(), start a new ServiceDiscovery() instance where the instantiation is embedded in AlluxioEtcdClient constructor, so in a way, AlluxioEtcdClient.connect gets called at time of instantiation.